pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/design/xml/ImmutableField.xml

```diff
@@ -924,4 +924,53 @@ class Foo {
             }
             ]]></code>
     </test-code>
+
+    <test-code>
+        <description>[java] ImmutableField should consider fields in anonymous classes #3956</description>
+        <expected-problems>2</expected-problems>
+        <expected-linenumbers>2,4</expected-linenumbers>
+        <code><![CDATA[
+public class C {
+    private int g = 0;  // can report a warning in this line
+    public Object o = new Object() {
+        private int f = 0;  // should report a warning in this line
+    };
+}
+]]></code>
+    </test-code>
+
+    <test-code>
+        <description>[java] ImmutableField should consider enum class #3957</description>
+        <expected-problems>1</expected-problems>
+        <expected-linenumbers>4</expected-linenumbers>
+        <code><![CDATA[
+public class D {
+    enum E {
+        R;
+        private int y = 42;  // should have a warning in this line
+    }
+}
+]]></code>
+    </test-code>
+
+    <test-code>
+        <description>#4046 [java] Incorrect ImmutableField error on field with both declaration and constructor initializer</description>
+        <expected-problems>0</expected-problems>
+        <code><![CDATA[
+public class NotSoImmutableField {
+    private int foo = 2;
+
+    NotSoImmutableField() {
+    }
+
+    NotSoImmutableField(final int foo) {
+        this.foo = foo;
+    }
+
+    public int getFoo() {
+        return foo;
+    }
+}
+]]></code>
+    </test-code>
```

Comment:
I don't think we should have this test… this is actually a FP (see #4869)

Reply:
On the other side, the discussion on #4869 is still open, whether the rule ImmutableField should be changed to report also such a case (which can't be easily solved by just adding `final`) or whether this case should be reported in a new rule.

Maybe we should for now just mention both issues here? #4046 and #4869 ?

In any case, this test "just" fixes, how the rule works at the moment - if we decide to solve #4869 in that way, to let ImmutableField report his case as well, this test case needs to be changed of course - but we'll notice it.

Reply:
I've added a comment to mention #4869 in this test case. I think, we sometimes have the case, that a rule improvement/bugfix invalidates older assumptions/test cases - and this would just be another example.
I hope, if this test fails, whoever is working on that, will see the comment and think a bit more about this (and come to the conclusion to either fix the test case or fix the rule).